packages/medusa/src/services/gift-card.ts

```diff
@@ -0,0 +1,280 @@
+          `Gift card is missing region_id`
+        )
+      }
+
+      // Will throw if region does not exist
+      const region = await this.regionService_.retrieve(giftCard.region_id)
+
+      const code = GiftCardService.generateCode()
+
+      const toCreate = {
+        code,
+        region_id: region.id,
+        ...giftCard,
+      }
+
+      const created = await giftCardRepo.create(toCreate)
+      const result = await giftCardRepo.save(created)
+
+      await this.eventBus_
+        .withTransaction(manager)
+        .emit(GiftCardService.Events.CREATED, {
+          id: result.id,
+        })
+
+      return result
+    })
+  }
+
+  protected async retrieve_(
+    selector: Selector<GiftCard>,
+    config: FindConfig<GiftCard> = {}
+  ): Promise<GiftCard> {
+    return await this.atomicPhase_(async (manager) => {
+      const giftCardRepo = manager.getCustomRepository(this.giftCardRepository_)
+
+      const { relations, ...query } = buildQuery(selector, config)
+
+      const giftCard = await giftCardRepo.findOneWithRelations(
+        relations as (keyof GiftCard)[],
+        query
+      )
+
+      if (!giftCard) {
+        const selectorConstraints = Object.entries(selector)
+          .map((key, value) => `${key}: ${value}`)
+          .join(", ")
+
+        throw new MedusaError(
+          MedusaError.Types.NOT_FOUND,
+          `Gift card with ${selectorConstraints} was not found`
+        )
+      }
+
+      return giftCard
+    })
+  }
+
+  /**
+   * Gets a gift card by id.
+   * @param giftCardId - id of gift card to retrieve
+   * @param config - optional values to include with gift card query
+   * @return the gift card
+   */
+  async retrieve(
+    giftCardId: string,
+    config: FindConfig<GiftCard> = {}
+  ): Promise<GiftCard> {
+    return await this.atomicPhase_(async () => {
+      return await this.retrieve_({ id: giftCardId }, config)
+    })
+  }
+
+  async retrieveByCode(
+    code: string,
+    config: FindConfig<GiftCard> = {}
+  ): Promise<GiftCard> {
+    return await this.atomicPhase_(async () => {
+      return await this.retrieve_({ code }, config)
+    })
+  }
+
+  /**
+   * Updates a giftCard.
+   * @param giftCardId - giftCard id of giftCard to update
+   * @param update - the data to update the giftCard with
+   * @return the result of the update operation
+   */
+  async update(
+    giftCardId: string,
+    update: UpdateGiftCardInput
+  ): Promise<GiftCard> {
+    return await this.atomicPhase_(async (manager) => {
+      const giftCardRepo = manager.getCustomRepository(this.giftCardRepository_)
+
+      const giftCard = await this.retrieve(giftCardId)
+
+      const { region_id, metadata, balance, ...rest } = update
+
+      if (region_id && region_id !== giftCard.region_id) {
+        const region = await this.regionService_.retrieve(region_id)
+        giftCard.region_id = region.id
+      }
+
+      if (metadata) {
+        giftCard.metadata = await setMetadata(giftCard, metadata)
```

Comment:
todo: Unnecessary `await`
```suggestion
        giftCard.metadata = setMetadata(giftCard, metadata)
```